components/dm/command/scale_in.go

```diff
@@ -64,3 +71,108 @@ func newScaleInCmd() *cobra.Command {
 
 	return cmd
 }
+
+// ScaleInDMCluster scale in dm cluster.
+func ScaleInDMCluster(
+	getter operator.ExecutorGetter,
+	spec *dm.Topology,
+	options operator.Options,
+) error {
+	// instances by uuid
+	instances := map[string]dm.Instance{}
+
+	// make sure all nodeIds exists in topology
+	for _, component := range spec.ComponentsByStartOrder() {
+		for _, instance := range component.Instances() {
+			instances[instance.ID()] = instance
+		}
+	}
+
+	// Clean components
+	deletedDiff := map[string][]dm.Instance{}
+	deletedNodes := set.NewStringSet(options.Nodes...)
+	for nodeID := range deletedNodes {
+		inst, found := instances[nodeID]
+		if !found {
+			return errors.Errorf("cannot find node id '%s' in topology", nodeID)
+		}
+		deletedDiff[inst.ComponentName()] = append(deletedDiff[inst.ComponentName()], inst)
+	}
+
+	// Cannot delete all DM DMMaster servers
+	if len(deletedDiff[dm.ComponentDMMaster]) == len(spec.Masters) {
+		return errors.New("cannot delete all dm-master servers")
+	}
+
+	if options.Force {
+		for _, component := range spec.ComponentsByStartOrder() {
+			for _, instance := range component.Instances() {
+				if !deletedNodes.Exist(instance.ID()) {
+					continue
+				}
+				// just try stop and destroy
+				if err := operator.StopComponent(getter, []dm.Instance{instance}, options.OptTimeout); err != nil {
+					log.Warnf("failed to stop %s: %v", component.Name(), err)
+				}
+				if err := operator.DestroyComponent(getter, []dm.Instance{instance}, spec, options); err != nil {
+					log.Warnf("failed to destroy %s: %v", component.Name(), err)
+				}
+
+				continue
+			}
+		}
+		return nil
+	}
+
+	// At least a DMMaster server exists
+	var dmMasterClient *api.DMMasterClient
+	var dmMasterEndpoint []string
+	for _, instance := range (&dm.DMMasterComponent{Topology: spec}).Instances() {
+		if !deletedNodes.Exist(instance.ID()) {
+			dmMasterEndpoint = append(dmMasterEndpoint, operator.Addr(instance))
+		}
+	}
+
+	if len(dmMasterEndpoint) == 0 {
+		return errors.New("cannot find available dm-master instance")
+	}
+
+	retryOpt := &utils.RetryOption{
+		Timeout: time.Second * time.Duration(options.APITimeout),
+		Delay:   time.Second * 2,
+	}
+	dmMasterClient = api.NewDMMasterClient(dmMasterEndpoint, 10*time.Second, nil)
+
+	// Delete member from cluster
+	for _, component := range spec.ComponentsByStartOrder() {
+		for _, instance := range component.Instances() {
+			if !deletedNodes.Exist(instance.ID()) {
+				continue
+			}
+
+			switch component.Name() {
+			case dm.ComponentDMMaster:
+				name := instance.(*dm.DMMasterInstance).Name
```

Comment:
Can we delete the prefix `DM` from type definition `DMMasterInstance` because it is duplicated with the `dm` package?